include/deal.II/base/patterns_tools.h

```diff
@@ -0,0 +1,682 @@
+    template <typename... Args> struct is_stl_container<std::priority_queue    <Args...>>:std::true_type {};
+
+    template <typename T>       struct is_stl_map:std::false_type {};
+    template <typename... Args> struct is_stl_map<std::map               <Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_map<std::multimap          <Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_map<std::unordered_map     <Args...>>:std::true_type {};
+    template <typename... Args> struct is_stl_map<std::unordered_multimap<Args...>>:std::true_type {};
+  }
+
+  //type trait to use the implementation type traits as well as decay the type
+  template <typename T> struct is_stl_container
+  {
+    static constexpr bool const value = internal::is_stl_container<std::decay_t<T>>::value;
+  };
+
+
+  template <typename T> struct is_stl_map
+  {
+    static constexpr bool const value = internal::is_stl_map<std::decay_t<T>>::value;
+  };
+
+
+  namespace internal
+  {
+    // Rank of vector types
+    template<class T>
+    struct RankInfo<T, typename std::enable_if<is_stl_container<T>::value>::type >
+    {
+      static constexpr int list_rank = RankInfo<typename T::value_type>::list_rank + 1;
+      static constexpr int map_rank = RankInfo<typename T::value_type>::map_rank;
+    };
+
+
+    // Rank of map types
+    template<class T>
+    struct RankInfo<T, typename std::enable_if<is_stl_map<T>::value>::type >
+    {
+      static constexpr int list_rank = std::max(RankInfo<typename T::key_type>::list_rank,
+                                                RankInfo<typename T::mapped_type>::list_rank) + 1;
+      static constexpr int map_rank = std::max(RankInfo<typename T::key_type>::map_rank,
+                                               RankInfo<typename T::mapped_type>::map_rank) + 1;
+    };
+
+    // Rank of Tensor types
+    template <int rank, int dim, class Number>
+    struct RankInfo<Tensor<rank, dim, Number>>
+    {
+      static constexpr int list_rank = rank + RankInfo<Number>::list_rank;
+      static constexpr int map_rank = RankInfo<Number>::map_rank;
+    };
+
+    template <int dim, class Number>
+    struct RankInfo<Point<dim,Number>>:RankInfo<Tensor<1,dim,Number>> {};
+
+    // Rank of complex types
+    template <class Number>
+    struct RankInfo<std::complex<Number>>
+    {
+      static constexpr int list_rank = RankInfo<Number>::list_rank+1;
+      static constexpr int map_rank = RankInfo<Number>::map_rank;
+    };
+  }
+
+
+  // stl containers
+  template<class T>
+  struct Convert<T, typename std::enable_if<is_stl_container<T>::value>::type>
+  {
+    static std::unique_ptr<PatternBase> to_pattern()
+    {
+      return std_cxx14::make_unique<List>(*Convert<typename T::value_type>::to_pattern(),
+                                          std::numeric_limits<typename T::size_type>::min(),
+                                          std::numeric_limits<typename T::size_type>::max(),
```

Comment:
And don't you get an overflow for the third argument? `Patterns::List` only takes `unsigned int`s in these two places.